File: src/main/resources/frontend/src/app/core/services/hotel-review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

/**
 * @interface HotelReview
 * @description Representa una reseña de hotel
 */
export interface HotelReview {
  id?: number;
  rating: number;
  title: string;
  comment: string;
  createdAt?: string;
  updatedAt?: string;
  editable?: boolean;
  
  // Información del usuario (limitada por privacidad)
  userId?: number; // ID del usuario para validaciones de permisos
  userFirstName?: string;
  userLastInitial?: string;
  
  // Información básica del hotel
  hotelId: number;
  hotelName?: string;
  
  // Información de la reserva
  bookingId: number;
  bookingDate?: string;
  checkInDate?: string;
  checkOutDate?: string;
  guestName?: string;
  reviewDate?: string;
}

/**
 * @interface CreateHotelReviewRequest
 * @description DTO para crear una nueva reseña
 */
export interface CreateHotelReviewRequest {
  bookingId: number;
  rating: number;
  title: string;
  comment: string;
}

/**
 * @interface UpdateHotelReviewRequest
 * @description DTO para actualizar una reseña existente
 */
export interface UpdateHotelReviewRequest {
  rating: number;
  title: string;
  comment: string;
}

/**
 * @interface HotelReviewStats
 * @description Estadísticas de reseñas de un hotel
 */
export interface HotelReviewStats {
  hotelId: number;
  hotelName?: string;
  totalReviews: number;
  averageRating: number;
  
  // Distribución de calificaciones
  oneStar: number;
  twoStars: number;
  threeStars: number;
  fourStars: number;
  fiveStars: number;
  
  // Porcentajes de cada calificación
  oneStarPercent: number;
  twoStarsPercent: number;
  threeStarsPercent: number;
  fourStarsPercent: number;
  fiveStarsPercent: number;
  
  // Compatibilidad - distribución como objeto
  ratingDistribution?: { [key: number]: number };
}

/**
 * @interface ReviewableBooking
 * @description Reserva que puede ser reseñada
 */
export interface ReviewableBooking {
  bookingId: number;
  confirmationCode: string;
  bookingDate: string;
  checkInDate: string;
  checkOutDate: string;
  hotelId: number;
  hotelName: string;
  hotelImageUrl?: string;
  hotelStars?: number;
  numGuests: number;
  numRooms: number;
  totalPrice: number;
}

/**
 * @interface PageResponse
 * @description Respuesta paginada
 */
export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

/**
 * @class HotelReviewService
 * @description Servicio para gestionar reseñas de hoteles
 */
@Injectable({
  providedIn: 'root'
})
export class HotelReviewService {
  private readonly API_URL = `${environment.apiUrl}/hotel-reviews`;

  constructor(private http: HttpClient) {}

  /**
   * Crea una nueva reseña de hotel
   * @param review Datos de la reseña
   * @returns Observable con la reseña creada
   */
  createReview(review: CreateHotelReviewRequest): Observable<HotelReview> {
    return this.http.post<HotelReview>(this.API_URL, review);
  }

  /**
   * Actualiza una reseña existente
   * @param reviewId ID de la reseña
   * @param review Datos actualizados
   * @returns Observable con la reseña actualizada
   */
  updateReview(reviewId: number, review: UpdateHotelReviewRequest): Observable<HotelReview> {
    return this.http.put<HotelReview>(`${this.API_URL}/${reviewId}`, review);
  }

  /**
   * Elimina una reseña
   * @param reviewId ID de la reseña
   * @returns Observable void
   */
  deleteReview(reviewId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${reviewId}`);
  }

  /**
   * Obtiene todas las reseñas de un hotel
   * @param hotelId ID del hotel
   * @param page Página
   * @param size Tamaño de página
   * @returns Observable con reseñas paginadas
   */
  getHotelReviews(hotelId: number, page: number = 0, size: number = 10): Observable<PageResponse<HotelReview>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<PageResponse<HotelReview>>(`${this.API_URL}/hotel/${hotelId}`, { params });
  }

  /**
   * Obtiene las reseñas de un usuario
   * @param userId ID del usuario
   * @param page Página
   * @param size Tamaño de página
   * @returns Observable con reseñas paginadas
   */
  getUserReviews(userId: number, page: number = 0, size: number = 10): Observable<PageResponse<HotelReview>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<PageResponse<HotelReview>>(`${this.API_URL}/user/${userId}`, { params });
  }

  /**
   * Obtiene estadísticas de reseñas de un hotel
   * @param hotelId ID del hotel
   * @returns Observable con estadísticas
   */
  getHotelReviewStats(hotelId: number): Observable<HotelReviewStats> {
    return this.http.get<HotelReviewStats>(`${this.API_URL}/hotel/${hotelId}/stats`);
  }

  /**
   * Obtiene reservas que pueden ser reseñadas por un usuario
   * @returns Observable con reservas reseñables
   */
  getReviewableBookings(): Observable<ReviewableBooking[]> {
    return this.http.get<ReviewableBooking[]>(`${this.API_URL}/reviewable-bookings`);
  }

  /**
   * Busca reseñas por texto
   * @param searchTerm Término de búsqueda
   * @param page Página
   * @param size Tamaño de página
   * @returns Observable con reseñas encontradas
   */
  searchReviews(searchTerm: string, page: number = 0, size: number = 10): Observable<PageResponse<HotelReview>> {
    const params = new HttpParams()
      .set('searchTerm', searchTerm)
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<PageResponse<HotelReview>>(`${this.API_URL}/search`, { params });
  }

  /**
   * Verifica si un usuario puede reseñar un hotel
   * @param hotelId ID del hotel
   * @returns Observable boolean
   */
  canUserReviewHotel(hotelId: number): Observable<boolean> {
    return this.http.get<boolean>(`${this.API_URL}/can-review/${hotelId}`);
  }

  /**
   * Obtiene las reseñas realizadas por un usuario específico
   * @param userId ID del usuario
   * @returns Observable con las reseñas del usuario
   */
  getReviewsByUser(userId: number): Observable<HotelReview[]> {
    return this.http.get<HotelReview[]>(`${this.API_URL}/user/${userId}`);
  }
}

File: src/main/java/org/project/caribevibes/service/hotel/HotelReviewService.java
package org.project.caribevibes.service.hotel;

import org.project.caribevibes.entity.booking.Booking;
import org.project.caribevibes.entity.hotel.HotelReview;
import org.project.caribevibes.repository.booking.BookingRepository;
import org.project.caribevibes.repository.hotel.HotelReviewRepository;
import org.project.caribevibes.exception.ResourceNotFoundException;
import org.project.caribevibes.exception.BusinessException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * Servicio para gestión de reseñas de hoteles.
 * 
 * @author Caribe Vibes Team
 * @version 1.0
 * @since 2025
 */
@Service
@Transactional
public class HotelReviewService {

    private static final Logger logger = LoggerFactory.getLogger(HotelReviewService.class);

    @Autowired
    private HotelReviewRepository reviewRepository;

    @Autowired
    private BookingRepository bookingRepository;

    /**
     * Crea una nueva reseña de hotel.
     * 
     * @param review Entidad de reseña a crear
     * @return Reseña creada
     * @throws BusinessException si ya existe una reseña para esta reserva
     */
    public HotelReview createReview(HotelReview review) {
        logger.info("Creando nueva reseña para hotel ID: {} por usuario ID: {}", 
                   review.getHotel().getId(), review.getUser().getId());

        // Verificar que no existe una reseña previa para esta reserva
        if (reviewRepository.existsByBookingIdAndUserId(review.getBooking().getId(), review.getUser().getId())) {
            throw new BusinessException("Ya existe una reseña para esta reserva");
        }

        // Verificar que la reserva esté completada o que haya pasado la fecha de check-out
        Booking booking = review.getBooking();
        boolean isReviewable = Booking.BookingStatus.COMPLETED.equals(booking.getStatus()) ||
                              (Booking.BookingStatus.CHECKED_OUT.equals(booking.getStatus()) && 
                               booking.getCheckOutDate().isBefore(LocalDate.now()));
        
        if (!isReviewable) {
            throw new BusinessException("Solo se pueden reseñar reservas completadas");
        }

        return reviewRepository.save(review);
    }

    /**
     * Actualiza una reseña existente.
     * 
     * @param reviewId ID de la reseña
     * @param userId ID del usuario (para validar permisos)
     * @param updatedReview Datos actualizados
     * @return Reseña actualizada
     * @throws ResourceNotFoundException si no se encuentra la reseña
     */
    public HotelReview updateReview(Long reviewId, Long userId, HotelReview updatedReview) {
        logger.info("Actualizando reseña ID: {} por usuario ID: {}", reviewId, userId);

        // SEGURIDAD: findByIdAndUserId garantiza que solo el propietario pueda actualizar su reseña
        HotelReview existingReview = reviewRepository.findByIdAndUserId(reviewId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Reseña", "id", reviewId));

        // Actualizar campos
        existingReview.setRating(updatedReview.getRating());
        existingReview.setTitle(updatedReview.getTitle());
        existingReview.setComment(updatedReview.getComment());

        return reviewRepository.save(existingReview);
    }

    /**
     * Elimina una reseña.
     * 
     * @param reviewId ID de la reseña
     * @param userId ID del usuario (para validar permisos)
     * @throws ResourceNotFoundException si no se encuentra la reseña
     */
    public void deleteReview(Long reviewId, Long userId) {
        logger.info("Eliminando reseña ID: {} por usuario ID: {}", reviewId, userId);

        // SEGURIDAD: findByIdAndUserId garantiza que solo el propietario pueda eliminar su reseña
        HotelReview review = reviewRepository.findByIdAndUserId(reviewId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Reseña", "id", reviewId));

        reviewRepository.delete(review);
    }

    /**
     * Obtiene todas las reseñas de un hotel.
     * 
     * @param hotelId ID del hotel
     * @param pageable Información de paginación
     * @return Página de reseñas
     */
    @Transactional(readOnly = true)
    public Page<HotelReview> getHotelReviews(Long hotelId, Pageable pageable) {
        logger.debug("Obteniendo reseñas para hotel ID: {}", hotelId);
        return reviewRepository.findByHotelIdOrderByCreatedAtDesc(hotelId, pageable);
    }

    /**
     * Obtiene las reseñas de un usuario.
     * 
     * @param userId ID del usuario
     * @param pageable Información de paginación
     * @return Página de reseñas
     */
    @Transactional(readOnly = true)
    public Page<HotelReview> getUserReviews(Long userId, Pageable pageable) {
        logger.debug("Obteniendo reseñas para usuario ID: {}", userId);
        return reviewRepository.findByUserIdOrderByCreatedAtDesc(userId, pageable);
    }

    /**
     * Obtiene las reseñas de un usuario (alias para getUserReviews).
     * 
     * @param userId ID del usuario
     * @param pageable Información de paginación
     * @return Página de reseñas
     */
    @Transactional(readOnly = true)
    public Page<HotelReview> getReviewsByUser(Long userId, Pageable pageable) {
        return getUserReviews(userId, pageable);
    }

    /**
     * Busca reseñas por texto.
     * 
     * @param searchTerm Término de búsqueda
     * @param pageable Información de paginación
     * @return Página de reseñas encontradas
     */
    @Transactional(readOnly = true)
    public Page<HotelReview> searchReviews(String searchTerm, Pageable pageable) {
        logger.debug("Buscando reseñas con término: {}", searchTerm);
        return reviewRepository.searchByTitleOrComment(searchTerm, pageable);
    }

    /**
     * Calcula el promedio de calificaciones de un hotel.
     * 
     * @param hotelId ID del hotel
     * @return Promedio de calificaciones
     */
    @Transactional(readOnly = true)
    public BigDecimal calculateHotelAverageRating(Long hotelId) {
        return reviewRepository.calculateAverageRatingByHotelId(hotelId);
    }

    /**
     * Cuenta el total de reseñas de un hotel.
     * 
     * @param hotelId ID del hotel
     * @return Número total de reseñas
     */
    @Transactional(readOnly = true)
    public Long countHotelReviews(Long hotelId) {
        return reviewRepository.countByHotelId(hotelId);
    }

    /**
     * Obtiene la distribución de calificaciones de un hotel.
     * 
     * @param hotelId ID del hotel
     * @return Array con conteos [1-star, 2-star, 3-star, 4-star, 5-star]
     */
    @Transactional(readOnly = true)
    public int[] getRatingDistribution(Long hotelId) {
        List<Object[]> distribution = reviewRepository.getRatingDistributionByHotelId(hotelId);
        int[] result = new int[5]; // Para calificaciones 1-5

        for (Object[] row : distribution) {
            Integer rating = (Integer) row[0];
            Long count = (Long) row[1];
            if (rating >= 1 && rating <= 5) {
                result[rating - 1] = count.intValue();
            }
        }

        return result;
    }

    /**
     * Obtiene las reservas que pueden ser reseñadas por un usuario.
     * 
     * @param userId ID del usuario
     * @return Lista de reservas reseñables
     */
    @Transactional(readOnly = true)
    public List<Booking> getReviewableBookings(Long userId) {
        logger.debug("Obteniendo reservas reseñables para usuario ID: {}", userId);
        
        // Buscar reservas completadas que no han sido reseñadas
        return bookingRepository.findCompletedBookingsWithoutReviews(userId);
    }

    /**
     * Verifica si un usuario puede reseñar un hotel específico.
     * 
     * @param userId ID del usuario
     * @param hotelId ID del hotel
     * @return true si puede reseñar
     */
    @Transactional(readOnly = true)
    public boolean canUserReviewHotel(Long userId, Long hotelId) {
        return reviewRepository.canUserReviewHotel(userId, hotelId);
    }

    /**
     * Obtiene reseñas recientes de un hotel.
     * 
     * @param hotelId ID del hotel
     * @param limit Número máximo de reseñas
     * @return Lista de reseñas recientes
     */
    @Transactional(readOnly = true)
    public List<HotelReview> getRecentHotelReviews(Long hotelId, int limit) {
        Pageable pageable = PageRequest.of(0, limit);
        return reviewRepository.findRecentReviewsByHotelId(hotelId, pageable);
    }
}
